File: web/src/validate.ts
// Client-side validators for the ACME form. These mirror the server checks in
// internal/core/validate.go — they exist only for instant feedback; the server
// is still the authority.

// isValidEmail reports whether s is a single, well-formed e-mail with a dotted
// domain part (e.g. "a@localhost" is rejected, as ACME CAs won't accept it).
export function isValidEmail(s: string): boolean {
  const v = s.trim()
  // local@domain, no spaces, exactly one @, domain has a dot and TLD ≥ 2 chars.
  return /^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/.test(v)
}

// isValidDomain reports whether s is a syntactically valid DNS hostname (FQDN
// with at least one dot, labels 1–63 chars of [A-Za-z0-9-], no leading/trailing
// hyphen). Not an IP — use isIP for that.
export function isValidDomain(s: string): boolean {
  const v = s.trim().replace(/\.$/, '')
  if (v.length === 0 || v.length > 253 || !v.includes('.')) return false
  const labels = v.split('.')
  const labelOk = labels.every((label) =>
    /^[A-Za-z0-9]([A-Za-z0-9-]{0,61}[A-Za-z0-9])?$/.test(label),
  )
  // The TLD (last label) is never all-numeric — also rejects IPv4 addresses,
  // which are syntactically valid label sequences.
  return labelOk && /[A-Za-z-]/.test(labels[labels.length - 1])
}

// isIP reports whether s is a valid IPv4 or IPv6 address.
export function isIP(s: string): boolean {
  const v = s.trim()
  // IPv4: four 0–255 octets.
  if (/^(\d{1,3}\.){3}\d{1,3}$/.test(v)) {
    return v.split('.').every((o) => Number(o) <= 255)
  }
  // IPv6: loose but practical — hex groups and colons, optional "::".
  return /^[0-9a-fA-F:]+$/.test(v) && v.includes(':')
}

// isValidACMETarget reports whether target is acceptable: Let's Encrypt accepts
// a domain OR an IP; ZeroSSL accepts domains only.
export function isValidACMETarget(target: string, zerossl: boolean): boolean {
  if (isValidDomain(target)) return true
  return !zerossl && isIP(target)
}

// RESERVED_SUB_PATHS mirrors reservedSubPaths in internal/core/manager_settings.go:
// first-segment names the subscription prefix must not use (panel/system routes).
const RESERVED_SUB_PATHS = new Set(['api', 'assets', 'login', 'logout', 'favicon', 'static', 'well-known'])

// subPathError validates the subscription path against the same rules as the
// server and returns an operator-facing error message, or null when valid.
// secret is the panel's secret path (the subscription must not shadow it).
export function subPathError(path: string, secret: string): string | null {
  const p = path.trim()
  if (!/^[A-Za-z0-9_-]{1,32}$/.test(p)) {
    return 'Путь подписки: латиница, цифры, «-» и «_», 1–32 символа.'
  }
  if (RESERVED_SUB_PATHS.has(p.toLowerCase())) {
    return `Путь «${p}» зарезервирован панелью — выберите другой.`
  }
  if (secret && p.toLowerCase() === secret.toLowerCase()) {
    return 'Путь подписки не может совпадать с секретным путём панели.'
  }
  return null
}

File: internal/core/validate.go
package core

import (
	"net"
	"net/mail"
	"strings"
)

// validEmail reports whether s is a single, well-formed e-mail address. It uses
// net/mail (RFC 5322) and additionally requires a dotted domain part so inputs
// like "a@localhost" are rejected — ACME CAs won't accept those.
func validEmail(s string) bool {
	s = strings.TrimSpace(s)
	addr, err := mail.ParseAddress(s)
	if err != nil || addr.Address != s {
		return false // reject display-name forms like "Name <a@b.com>"
	}
	at := strings.LastIndexByte(s, '@')
	domain := s[at+1:]
	return strings.Contains(domain, ".") && !strings.HasPrefix(domain, ".") &&
		!strings.HasSuffix(domain, ".")
}

// validDomain reports whether s is a syntactically valid DNS hostname (a FQDN
// with at least one dot, each label 1–63 chars of [A-Za-z0-9-], not starting or
// ending with a hyphen). It is intentionally NOT an IP — callers test IPs
// separately with net.ParseIP.
func validDomain(s string) bool {
	s = strings.TrimSpace(strings.TrimSuffix(s, "."))
	if len(s) == 0 || len(s) > 253 || !strings.Contains(s, ".") {
		return false
	}
	labels := strings.Split(s, ".")
	for _, label := range labels {
		if len(label) == 0 || len(label) > 63 {
			return false
		}
		if label[0] == '-' || label[len(label)-1] == '-' {
			return false
		}
		for i := 0; i < len(label); i++ {
			c := label[i]
			if !(c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' || c >= '0' && c <= '9' || c == '-') {
				return false
			}
		}
	}
	// The TLD (last label) is never all-numeric — this also rejects IPv4
	// addresses, which are syntactically valid label sequences.
	tld := labels[len(labels)-1]
	for i := 0; i < len(tld); i++ {
		if tld[i] < '0' || tld[i] > '9' {
			return true
		}
	}
	return false
}

// NormalizeACMEHost lowercases domain targets for ACME (CAs canonicalize DNS
// names; mixed case like WaifuVPN.example.com makes lego reject the order).
func NormalizeACMEHost(target string) string {
	target = strings.TrimSpace(strings.TrimSuffix(target, "."))
	if validDomain(target) {
		return strings.ToLower(target)
	}
	return target // IP addresses and odd inputs pass through unchanged
}

// validACMETarget reports whether target is acceptable for the given provider:
// Let's Encrypt accepts a domain OR an IP; ZeroSSL accepts domains only.
func validACMETarget(target, provider string) bool {
	if validDomain(target) {
		return true
	}
	if provider != "zerossl" && net.ParseIP(target) != nil {
		return true
	}
	return false
}
